azure-iot-device/tests/common/transport/test_pipeline_stages_base.py

```diff
@@ -0,0 +1,544 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import pytest
+import functools
+from azure.iot.device.common.transport import pipeline_stages_base
+from azure.iot.device.common.transport import pipeline_ops_base
+from azure.iot.device.common.transport import pipeline_events_base
+
+logging.basicConfig(level=logging.INFO)
+
+
+def assert_default_attributes(obj):
+    assert obj.name is obj.__class__.__name__
+    assert obj.next is None
+    assert obj.previous is None
+    assert obj.pipeline_root is None
+
+
+# because PipelineStage is abstract, we need something concrete
+class PipelineStage(pipeline_stages_base.PipelineStage):
+    def _run_op(self, op):
+        pass
```

Comment:
I know that the EnsureConnection tests aren't in yet, but just as an idea, have you considered parametrizing some of these pipeline stage tests so that they will test all subclasses of PipelineStage as well?

After all, the basics (and complexity) of how a PipelineStage works is inherited, so a more specific PipelineStage should pass all the tests the generic PipelineStage does, right?

Instead of hacking a PipelineStage object to instantiate even though it's an AbstractBaseClass, why not ONLY test the child classes? After all, the parent class is NEVER instantiated in practice as it is in this test file, it's only actually done through the children. By parametrizing your objects under test among the various PipelineStage child classes, you can use one set of tests to cover all PipelineStage children.

This accomplishes 2 things:

1) It tests the classes as they are actually used in practice.
2) It gives us assurance that these tests pass for all child classes. 

After all, testing the parent class only tells us that those rules are valid for the children by convention - we take for granted the children don't muck it up somehow.

I hope that all makes sense.

Reply:
That does make sense.  I can see your argument, but I'm not sure if I agree.  I think that kind of testing _could_ have value, and it could save us some copy/paste in our tests, but I'm not sure how much value it would actually add.  

I think you're asking us to really think hard about what the word "unit" in "unit tests" means.  We usually define "unit" as a single piece of code (an object or a function), and we mock everything outside of that unit.  So, what is the "unit" we are testing here?  

As a thought exercise, think about using composition instead of inheritance for these objects and how we would test that?  In other words, if an Mqtt pipeline stage _contained_ a base pipeline object to do the base work (a "has-a" relationship, as in "an MQTT stage has a base object that it can use to do work"), it's _almost_ exactly the same as our case where an mqtt pipeline stage inherits the behavior of the the base stage (an "is-a" relationship, as in "an MQTT stage is-a base stage").  In the "has-a" example, we would mock the base object and only test that the MQTT object called into the base object.  We wouldn't actually test the base object.  Why is this different?

Another way to think about it is to think about how waterproof our assurances can be is to think about how we test the derived classes.  

First of all, if a derived object never calls run_ops_serial on itself, then we don't get anything extra by verifying that run_ops_serial works correctly in the derived class.  Sure, we've verified that this function is available in the derived class and that it behaves correctly, but it's never called, so who cares?   

However, if a derived class uses run_ops_serial, it's using it to do something specific, and that specific thing is part of the derived class' requirements.  For example, if a derived class uses run_ops_serial to run 3 ops in a row, our derived class tests had better verify the behavior in running those 3 ops.  (e.g. the first one runs, if the first one succeeds, the second one runs, etc.). The requirements on that derived class should specify that the 3 ops run in a row, and it should specify how to handle errors and whatnot, and the derived class tests should validate that.  The fact that the derived class uses run_ops_serial is an implementation detail.  

Testing the full range of run_ops_serial does not give us any advantage here either because we only really care about the subset of functionality that the derived class actually uses.  This is the same as the "derived class never calls run_ops_serial" case, except it's more nuanced.  If we verify that there's a run_ops_serial function in the derived class that works correctly with some set of parameters, but the derived class never actually calls it with those parameters, what value do we get in testing the derived class with those parameters?

To draw a visual analogy, say we're testing a button on some screen and that button has some text on it.  The base implementation for that button also supports buttons with pictures on them.  Do we need to test how this button will behave with a picture on it, even if this button will never have a picture on it?  Sure, we could, but it's a slippery slope.  

---
In reply to: [282134493](https://github.com/Azure/azure-iot-sdk-python-preview/pull/89#discussion_r282134493) [](ancestors = 282134493)